mantle/kola/tests/misc/network.go

```diff
@@ -219,6 +228,67 @@ var (
           umount /var/mnt
 	`
 
+	setupVethPairsTemplate = `#!/usr/bin/env bash
+          set -ex
+
+          create_veth_pair() {
+            veth_host_side_end_name=$1
+            veth_netns_side_end_name=$2
+            host_side_mac_address=$3
+            netns_side_ip=$4
+            network_namespace=$5
+
+            # Create veth pair and assign a namespace to veth-netns
+            ip link add ${veth_host_side_end_name} type veth peer name ${veth_netns_side_end_name}
+            ip link set ${veth_netns_side_end_name} netns ${network_namespace}
+
+            # Assign an MAC address to the host-side veth end
+            ip link set dev ${veth_host_side_end_name} address ${host_side_mac_address}
+
+            # Assign an IP address to netns-side veth end and bring it up
+            ip netns exec ${network_namespace} ip address add ${netns_side_ip} dev ${veth_netns_side_end_name}
+            ip netns exec ${network_namespace} ip -6 addr flush ${veth_netns_side_end_name}
+            ip netns exec ${network_namespace} ip link set ${veth_netns_side_end_name} up
+          }
+
+          activate_veth_end() {
+            veth_end_name=$1
+            nmcli dev set ${veth_end_name} managed yes
+            ip link set ${veth_end_name} up
+          }
+
+          primary_mac=%s
+          secondary_mac=%s
+          primary_ip=%s
+          secondary_ip=%s
+
+          network_namespace=test-netns
+          # Create a network namespace
+          ip netns add ${network_namespace}
+
+          # create 2 veth pairs between host side and network_namespace
+          veth1_host_side_end_name=veth1-host
+          veth1_netns_side_end_name=veth1-netns
+          create_veth_pair ${veth1_host_side_end_name} ${veth1_netns_side_end_name} ${primary_mac} 192.168.0.59 ${network_namespace}
+          veth2_host_side_end_name=veth2-host
+          veth2_netns_side_end_name=veth2-netns
+          create_veth_pair ${veth2_host_side_end_name} ${veth2_netns_side_end_name} ${secondary_mac} 192.168.0.60 ${network_namespace}
+
+          # Run a dnsmasq service on the network_namespace, to set the host-side veth ends a ip via their MAC addresses
+          echo -e "dhcp-range=192.168.0.50,192.168.0.60,255.255.255.0,12h\ndhcp-host=${primary_mac},${primary_ip}\ndhcp-host=${secondary_mac},${secondary_ip}" > /etc/dnsmasq.d/dhcp
+          ip netns exec ${network_namespace} dnsmasq --log-queries --log-facility=/var/log/dnsmasq.log &
+
+          # Tell NM to manage the "veth-host" interface and bring it up (will attempt DHCP).
+          # Do this after we start dnsmasq so we don't have to deal with DHCP timeouts.
+          activate_veth_end ${veth1_host_side_end_name}
+          activate_veth_end ${veth2_host_side_end_name}
+          # Wait to let the dhcp request to be processed
+          sleep 10
```

Comment:
Could we use polling for the profile to be activated instead? The sleep could lead to races

Reply:
sure. Done